### DISCUSSION TITLE
Is average precision calculated using a binary mask for each class or binary mask for each instance of each class ?

### DISCUSSION THREAD

**Discussion by ExtReMLapin** [ultralytics/yolov5#9838] [Q&A] [answered]
Hello,

As the title states, i'm wondering if the calculation of the average precision (and mAP and mAP 0.5) is done the following : 

During training : 

`for each predicted classes, make a binary mask of the size of the image and calulate IoU.`
Or is it :
`for each predicted instance of each classes, check if there is any compatible match in the validation set and calculated IoU based on this`

**The difference of the two approaches could make a difference if the model predicts two instances at the same-ish position but for only one instance in the validation set, which means it would turn a IoU of 1 for the first approach to a IoU of `2/3` for the second approach.**

Let me explain you why i'm wondering this.

Right now I have some issues having good metrics.
![image](https://user-images.githubusercontent.com/3909752/196362156-faf3786f-870c-4c93-8b10-b0dbef1358d1.png)
I gave a try with different architectures, (one from metrics is yolov5n6) : 
* P2
* P34
* yolov5l6

However, I realized that by artifially increasing the size of each annotated text, it increases my mAP score.

Obviously, if the implementation is the first proposed one, it means it just takes false positives and turn them into true positives (as the annotations size in val set were increased) , and possibly false negative into true positive (as the model now predicts larger objects, even if predicted incorrectly).

However, if the implementation is the second one, it should only increases if it gives to the AI model a better context of the object.
I have a single-class with 15k annotations and I fear the expert who annotated it sometimes drew a too small rect, this is why i ran this experiment that gave me better results.

![image](https://user-images.githubusercontent.com/3909752/196363546-a8fef1fc-4cfc-4e73-a9fe-09600011cbf6.png)

Thank you for your time.

**Comment by ExtReMLapin** [comment]
In other words : 

During evaluation of what comes to the result of mAP

Does it : 

Solution 1:

Take one binary mask of the expected results
One binary mask of the predicted result
Calculate IoU using those two masks.

(Count intersection pixels, non matching and get IoU out of it)

Or : 
does it do something like that

```pseudocode
IoU_array_cur_class = []
for each predicted_box:
   if there is any overlapping val_box:
      take the val_box with the highest IoU and remove the current predicted_box and remove val_box (so it can't be matched again)
      cur_IoU = IoU(val_box, predicted_box)
   else if there is no overlapping box:
      cur_IoU = 0
  IoU_array_cur_class.append(cur_IoU)
```

**Comment by spacewalk01** [comment, marked as answer]
IoU is calculated by comparing the predicted box and ground truth box, (not mask).

**Reply by ExtReMLapin** [reply]
Thanks
